### PR TITLE
add base structure and serialization feature

### PR DESCRIPTION
**Pull request by kafeikui**

### CODE REVIEW

File: src/model/config.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+export class Snapshot {
+    id: string;
+    nonce: number;
+    baseSnapshot: Snapshot;
+    poolConfig: PoolConfig;
+    records: Record[];
+    baseRecordId: string;
+    lastRecordId: string;
+    used: boolean;
```
Comment (neptune-v) [inline src/model/config.ts:28]: For this kind of data record structure, ideally, we should keep them immutable and stateless after being retrieved from the database. Adding state-tracking fields/properties tends to mislead developers to treat these objects as mutable and update non-state-tracking fields when working on the code.

This rule should apply to other data objects as well.

We could track any possible state in the higher-level code that is consuming the objects.

File: src/model/config.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    tickSpacing: number;
+    token0: string;
+    token1: string;
+    fee: number;
+}
+
+export class Record {
+    id: string;
+    eventId: number;
+    actionName: string;
```
Comment (neptune-v) [inline src/model/config.ts:15]: Considering the types of actions are limited and not arbitrary, maybe it's safer and more clear if using a typed `enum` e.g. `ActionType` here and updating the variable name to `actionType`.

File: src/model/config.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+export class PoolConfig {
+    id: string;
+    tickSpacing: number;
+    token0: string;
+    token1: string;
+    fee: number;
+}
+
+export class Record {
```
Comment (neptune-v) [inline src/model/config.ts:12]: Is `Record` equivalent to `Event`?
If so, any reason that we are not just calling it `Event`?

File: src/model/config.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+export class SnapshotStorage {
+    token0Balance: BN;
+    token1Balance: BN;
+    sqrtPriceX96: BN;
+    tickCurrent: number
+    feeGrowthGlobal0X128: BN;
+    feeGrowthGlobal1X128: BN;
+    ticks_json: object;
```
Comment (neptune-v) [inline src/model/config.ts:51]: If we have to save the binary data of an object into a database, maybe SQL is not the best option.
I am not sure about the upper limit of the number of ticks or positions, but the better practice is definitely using a No-SQL database for directly storing vertically growing objects.

**Review by neptune-v — CHANGES_REQUESTED** [review]
Great job outlining the architecture and components!!!

I left some comments and questions.
Just to clarify, I am being particularly picky on the initial code review.
Problems or concerns I might have for the details do not diminish the efforts or the code's value!

Can we also add `[Prettier](https://prettier.io/)` for linting?

**Comment by neptune-v** [comment]
Initial review completed, please take a look.

**Comment by neptune-v** [comment]
The initial commit is an exception, but I want to emphasize that for future PRs, let's keep the changeset small and manageable.

### Each PR should only contain **one** single feature or **one** issue resolution.

The goal for this is to make it easier for the code reviewer to read all of the code and fully understand what the code does then make constructive comments to improve the code quality as a combined effort. 

If the changeset is too big, the reviewer will get lost in the complicated logic when tracking multiple components at the same time, then the reviewer will miss things, getting lazy not trying to understand everything, eventually the code-review process would become merely a formality and a burden for everyone without providing values as to improve the quality and transparency of the code base.

File: src/model/config.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    private persist(configurableCorePool: ConfigurableCorePool): string {
+        // TODO
+        return '0';
+    }
+
+}
+
+export class SnapshotStorage {
```
Comment (neptune-v) [inline src/model/config.ts:44]: I looked at the SQL table for Snapshot, maybe we should just flatten the `Snapshot` so it contains all the actual data currently stored in `SnapshotStorage`?

File: src/model/config.ts [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    eventId: number;
+    actionName: string;
+    actionParams: object;
+    timestamp: Date;
+}
+
+export class Snapshot {
+    id: string;
+    nonce: number;
+    baseSnapshot: Snapshot;
```
Comment (neptune-v) [inline src/model/config.ts:23]: If the `Snapshot` is not incremental, and it already contains all the records, we probably do not need this field then?

File: src/model/decorator.ts [hunk 7]
```diff
@@ -0,0 +1,8 @@
+import "reflect-metadata";
```
[the remark this answers is not in the document]
Comment (kafeikui) [reply src/model/decorator.ts:1]: It is for defining and reading metadata on types. I use it to support `Serialize` decorator so that every class can customize which property to be serialized.

File: src/core/pool.ts [hunk 8]
```diff
@@ -0,0 +1,78 @@
+import BN from 'bn.js';
+import { PositionManager, TickManager } from '../managers/pool';
+import { Tick, Position } from '../model/pool';
+import { TickMath } from '../utils/math';
+
+export class CorePool {
```
[the remark this answers is not in the document]
Comment (kafeikui) [reply src/core/pool.ts:6]: Actually I'm not sure name convention of typescript. If we take a file as a module, then it is able to contain sereval classes?

File: src/client/cli.ts [hunk 9]
```diff
@@ -0,0 +1,24 @@
+import { ConfigurableCorePool } from "../core/configurablePool";
+import { PoolConfig, Snapshot } from "../model/config";
+
+export class SimulatorClient {
```
[the remark this answers is not in the document]
Comment (kafeikui) [reply src/client/cli.ts:4]: not for the CLI, this is the library's entrance, an api for the library's user.

File: src/managers/pool.ts [hunk 10]
```diff
@@ -0,0 +1,52 @@
+import BN from "bn.js";
```
[the remark this answers is not in the document]
Comment (kafeikui) [reply src/managers/pool.ts:1]: sounds good.

File: tsconfig.json [hunk 11]
```diff
@@ -0,0 +1,12 @@
+{
```
[the remark this answers is not in the document]
Comment (kafeikui) [reply tsconfig.json:1]: `noImplicitAny` rule from strict mode is not friendly to implementation of serialization now. I will see what i can do to make it fit the rule.

File: src/model/config.ts [hunk 12]
```diff
@@ -0,0 +1,54 @@
+import BN from "bn.js";
+import { ConfigurableCorePool } from "../core/configurablePool";
+
+export class PoolConfig {
```
[the remark this answers is not in the document]
Comment (kafeikui) [reply src/model/config.ts:4]: I don't know `type` actually, I think it's more proper.

File: src/core/configurablePool.ts [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+            let states = snapshot.snapshotStorage;
+            super(poolConfig.token0, poolConfig.token1, poolConfig.fee, poolConfig.tickSpacing,
+                states.token0Balance, states.token1Balance, states.sqrtPriceX96, states.tickCurrent,
+                states.feeGrowthGlobal0X128, states.feeGrowthGlobal1X128,
+                new TickManager().fromJSON(states.ticks_json), new PositionManager().fromJSON(states.positions_json));
+            this._snapshot = snapshot;
+        }
+    }
+
+    mint(tickLower: number, tickUpper: number, amount: BN) {
```
[the remark this answers is not in the document]
Comment (kafeikui) [reply src/core/configurablePool.ts:24]: As far as I know typescript doesn't require `function` keyword with a class method.

File: src/model/element.ts [hunk 14]
```diff
@@ -0,0 +1,45 @@
+import "reflect-metadata";
```
[the remark this answers is not in the document]
Comment (kafeikui) [reply src/model/element.ts:1]: I will take a look.

File: src/core/configurablePool.ts [hunk 15]
```diff
@@ -0,0 +1,64 @@
+import BN from "bn.js";
+import { PositionManager, TickManager } from "../managers/pool";
+import { PoolConfig, Snapshot, Record } from '../model/config';
+import { CorePool } from './pool'
+
+export class ConfigurableCorePool extends CorePool {
+    private _snapshot: Snapshot;
```
[the remark this answers is not in the document]
Comment (kafeikui) [reply src/core/configurablePool.ts:7]: I see. Because there is a name collision between the private property and a function below. will rename the function.

File: src/model/config.ts [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+
+export class Record {
+    id: string;
+    eventId: number;
+    actionName: string;
+    actionParams: object;
+    timestamp: Date;
+}
+
+export class Snapshot {
```
[the remark this answers is not in the document]
Comment (kafeikui) [reply src/model/config.ts:20]: So far I think `Snapshot` is more like metadata to tell how it comes there, and `SnapshotStorage` is more like a state of the pool at some time if we choose to take a snapshot.
Singular is more accurate, I think.

File: src/model/config.ts [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+    id: string;
+    nonce: number;
+    baseSnapshot: Snapshot;
+    poolConfig: PoolConfig;
+    records: Record[];
+    baseRecordId: string;
```
[the remark this answers is not in the document]
Comment (kafeikui) [reply src/model/config.ts:26]: yes they are reserved for use I can't tell clearly. Agree to delete them for now.

File: src/model/config.ts (hunk 3 above)
Comment (kafeikui) [reply src/model/config.ts:12]: both OK. I kind of want to distinguish it from event emitted by a contract calling, since these events is also in the db now.

File: src/model/config.ts (hunk 4 above)
Comment (kafeikui) [reply src/model/config.ts:51]: will think about it.

File: src/model/config.ts (hunk 1 above)
Comment (kafeikui) [reply src/model/config.ts:28]: There is a state structure immutable from the database, and also a structure tracking current simulator's running history. It seems now they are kind of mixed.

File: src/model/config.ts (hunk 2 above)
Comment (kafeikui) [reply src/model/config.ts:15]: I agree.

File: src/model/config.ts (hunk 6 above)
Comment (kafeikui) [reply src/model/config.ts:23]: This is because I want to track how a snapshot came to state it is now from initial deployment action.

File: src/model/config.ts (hunk 5 above)
Comment (kafeikui) [reply src/model/config.ts:44]: It's also mixture problem mentioned above. looks like it's quite confusing. will think about it.

File: src/core/pool.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (neptune-v) [reply src/core/pool.ts:6]: True, there is no official style guide as to how the project should be organized.
Let's use the convention that we are familiar with, in Java and C#, let's use the filename to reflect the classname directly.

File: src/client/cli.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (neptune-v) [reply src/client/cli.ts:4]: Let's update the filename then if it's not a CLI.

File: src/core/configurablePool.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (neptune-v) [reply src/core/configurablePool.ts:24]: Let's see if ts compiler likes it in the strict mode and if `prettier` complains about it.
From my observation of the official TypeScript code snippets in the docs and other projects, they do keep the `function` keywords.

File: src/core/configurablePool.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (neptune-v) [reply src/core/configurablePool.ts:7]: Yeah, maybe `takeSnapshot` or something

File: src/model/config.ts (hunk 1 above)
Comment (neptune-v) [reply src/model/config.ts:28]: Just to clarify, by "state", I meant in the sense of general programming, like "a state of an object", not the "state" in the simulator.

As a pure data object, these structures shouldn't have a mutable internal state, it's better to make them plain, immutable objects that are being passed around.

File: src/model/config.ts (hunk 3 above)
Comment (neptune-v) [reply src/model/config.ts:12]: I see, makes sense, `record` is good then.

File: src/model/config.ts (hunk 6 above)
Comment (neptune-v) [reply src/model/config.ts:23]: I see what you are saying, maybe do something like this:

1. Add another class called `PoolState` which represents the current state of the pool.

2. Keep the class `Snapshot`, but flatten it so it contains everything in the `SnapshotStorage`, and remove class `SnapshotStorage`.

3. In the `PoolState`, it will have `baseSnapshot`  to keep track of which snapshot that the current state is recovered from, it will also have a list of `records` (deltas) that have been applied on top of the `baseSnapshot` to achieve the current state.

In short, `baseSnapshot` + `records` = `PoolState`

File: src/model/config.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (neptune-v) [reply src/model/config.ts:20]: See below, it might resolve the confusion.

File: src/model/decorator.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (neptune-v) [reply src/model/decorator.ts:1]: I see, nonetheless, if `TypedJSON` can satisfy our need, it might be more robust than handling the serialization/deserialization on our own.

File: src/core/configurablePool.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (kafeikui) [reply src/core/configurablePool.ts:24]: will do. I also see a lot of use as `export function` or `function` in a ts file but not in a scope of class. I think maybe it's for distinction with method of a class.

File: src/model/config.ts (hunk 6 above)
Comment (kafeikui) [reply src/model/config.ts:23]: I like this solution. Thank you.

File: tsconfig.json (hunk 11 above)
[the remark this answers is not in the document]
Comment (kafeikui) [reply tsconfig.json:1]: * "forceConsistentCasingInFileNames": true

File: src/model/Position.ts [hunk 18]
```diff
@@ -0,0 +1,17 @@
+import BN from "bn.js";
+
+export class Position {
+  liquidity: BN = new BN(0);
+  feeGrowthInside0LastX128: BN = new BN(0);
+  feeGrowthInside1LastX128: BN = new BN(0);
+  tokensOwed0: BN = new BN(0);
+  tokensOwed1: BN = new BN(0);
+
+  update(
```
Comment (neptune-v) [inline src/model/Position.ts:10]: `update` mutates the internal data of the data structure, yet the fields that are being updated are also writable directly e.g. `position.feeGrowthInside0LastX128 = <some new value>`, this can cause inconsistency (updated via `update` v.s. updated directly), consider making the fields that are updated by `update` private, then make them read-only by adding getters.

File: src/model/Tick.ts [hunk 19]
```diff
@@ -0,0 +1,25 @@
+import BN from "bn.js";
+
+export class Tick {
+  liquidityGross: BN = new BN(0);
+  liquidityNet: BN = new BN(0);
+  feeGrowthOutside0X128: BN = new BN(0);
+  feeGrowthOutside1X128: BN = new BN(0);
+  initialized: boolean = false;
+
+  update(
```
Comment (neptune-v) [inline src/model/Tick.ts:10]: ditto

File: src/model/PoolState.ts [hunk 20]
```diff
@@ -0,0 +1,38 @@
+import { ConfigurableCorePool } from "../core/ConfigurableCorePool";
+import { PoolConfig } from "../entity/PoolConfig";
+import { Record } from "../entity/Record";
+import { Snapshot } from "../entity/Snapshot";
+import { DBManager } from "../manager/DBManager";
+
+export class PoolState {
+  id: string = "0";
+  baseSnapshot: Snapshot | null;
+  snapshot: Snapshot | null = null;
```
Comment (neptune-v) [inline src/model/PoolState.ts:10]: I read that `undefined` is preferred to `null`.

See: https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html#use-optional-parameters

File: src/core/ConfigurableCorePool.ts [hunk 21]
```diff
@@ -0,0 +1,126 @@
...
+  takeSnapshot(): string {
+    // TODO
+    return "0";
+  }
+
+  recover(snapshotId: string) {
+    // TODO
+  }
+
+  private doRecord(actionType: ActionType, actionParams: object): Record {
```
Comment (neptune-v) [inline src/core/ConfigurableCorePool.ts:107]: Consider renaming the function to `applyRecord`

File: src/manager/PositionManager.ts [hunk 22]
```diff
@@ -0,0 +1,29 @@
...
+
+  set(owner: string, tickLower: number, tickUpper: number) {
+    // TODO
+  }
+
+  setKey(key: string, position: Position) {
+    this.positions.set(key, position);
+  }
+
+  getKey(key: string): Position {
```
Comment (neptune-v) [inline src/manager/PositionManager.ts:26]: The name is `getKey`, but it returns a `Position`

File: src/manager/TickManager.ts [hunk 23]
```diff
@@ -0,0 +1,46 @@
...
+
+  set(tickIndex: number, tick: Tick) {
+    this.ticks.set(tickIndex, tick);
+  }
+
+  flipTick(tickIndex: number, tickSpacing: number) {
+    // TODO
+  }
+
+  nextInitializedTick(
```
Comment (neptune-v) [inline src/manager/TickManager.ts:26]: Use verb in all functions, maybe pre-pend `add`.

File: src/model/PoolState.ts [hunk 24]
```diff
@@ -0,0 +1,38 @@
...
+  poolConfig: PoolConfig;
+  records: Record[] = new Array<Record>();
+  timestamp: Date = new Date();
+
+  constructor(baseSnapshot: Snapshot | null = null, poolConfig: PoolConfig) {
```
Comment (neptune-v) [inline src/model/PoolState.ts:15]: Consider using the `optional parameters`, see if it fits the need.

File: src/core/ConfigurableCorePool.ts [hunk 25]
```diff
@@ -0,0 +1,126 @@
...
+    actionType: ActionType,
+    actionParams: object
+  ) => void = function (
+    this: ConfigurableCorePool,
+    recordId: string,
+    actionType: ActionType,
+    actionParams: object
+  ) {};
+
+  constructor(poolConfig: PoolConfig);
```
Comment (neptune-v) [inline src/core/ConfigurableCorePool.ts:26]: Consider using the `optional parameters` for constructor overloading.

**Review by neptune-v — CHANGES_REQUESTED** [review]
Please see the inline comments.

File: src/model/Position.ts (hunk 18 above)
Comment (kafeikui) [reply src/model/Position.ts:10]: ok

File: src/model/Tick.ts (hunk 19 above)
Comment (kafeikui) [reply src/model/Tick.ts:10]: ok

File: src/manager/PositionManager.ts (hunk 22 above)
Comment (kafeikui) [reply src/manager/PositionManager.ts:26]: ok

File: src/core/ConfigurableCorePool.ts (hunk 21 above)
Comment (kafeikui) [reply src/core/ConfigurableCorePool.ts:107]: Just to make sure, does `apply` here means to write down(the record)?

File: src/model/PoolState.ts (hunk 20 above)
Comment (kafeikui) [reply src/model/PoolState.ts:10]: Could you please tell more about why undefined is preferred to null in this section?

File: src/manager/TickManager.ts (hunk 23 above)
Comment (kafeikui) [reply src/manager/TickManager.ts:26]: I see and I think `pick` as pre-pend is better here.

File: src/model/PoolState.ts (hunk 24 above)
Comment (kafeikui) [reply src/model/PoolState.ts:15]: I see this point. And actually here I want to choose one of two params as the input to construct the instance. I wonder if there is some good idea.

File: src/model/PoolState.ts (hunk 20 above)
Comment (kafeikui) [reply src/model/PoolState.ts:10]: Forget it, I think I see your point.

File: src/manager/TickManager.ts (hunk 23 above)
Comment (neptune-v) [reply src/manager/TickManager.ts:26]: I meant to say `get`

File: src/model/PoolState.ts (hunk 24 above)
Comment (neptune-v) [reply src/model/PoolState.ts:15]: I see, which one do you want to use with higher priority?
I think if both are on the same level of priority, it's actually better just to have 2 constructors, one for each, instead of a `if` switch, constructor apis like this are a little confusing.

File: src/core/ConfigurableCorePool.ts (hunk 21 above)
Comment (neptune-v) [reply src/core/ConfigurableCorePool.ts:107]: Nope, and me misunderstanding the meaning of the function is the problem caused by the naming.
In this case, `PersistRecord` might be more appropriate.

File: src/model/PoolState.ts (hunk 20 above)
Comment (neptune-v) [reply src/model/PoolState.ts:10]: https://stackoverflow.com/questions/5076944/what-is-the-difference-between-null-and-undefined-in-javascript

File: src/core/ConfigurableCorePool.ts (hunk 25 above)
Comment (neptune-v) [reply src/core/ConfigurableCorePool.ts:26]: Sorry, I stand corrected, this wasn't overloading, please ignore my comment.

**Review by neptune-v — APPROVED** [review]
Once these changes are updates, you can merge the PR.

File: src/model/PoolState.ts (hunk 24 above)
Comment (kafeikui) [reply src/model/PoolState.ts:15]: Actually typescript doesn't support multiple constructors and it supports constructor overloading like multiple signatures and one implementation. But I think (param1? : type1, param2? : type2) will work. Please have a look after next commit.

File: src/core/ConfigurableCorePool.ts (hunk 25 above)
Comment (kafeikui) [reply src/core/ConfigurableCorePool.ts:26]: No, you are correct. It's overloading but not what I want actually.

File: src/model/PoolState.ts (hunk 20 above)
Comment (kafeikui) [reply src/model/PoolState.ts:10]: well I think null is better to use here for meaning part and undefined is acceptable for its practice.

File: src/core/ConfigurableCorePool.ts (hunk 21 above)
Comment (kafeikui) [reply src/core/ConfigurableCorePool.ts:107]: Let me rephrase, I kind of think `apply` is just fine because this function is to build a record then append it to the current `PoolState`. Do we share the same idea?

Comment (neptune-v) [reply src/core/ConfigurableCorePool.ts]: Oh then my original understanding was correct, `applyRecord` means applying the delta on top of a state.

**Review by neptune-v — APPROVED** [review]
LGTM, thanks for the effort!!
Please merge.
